src/Microsoft.Azure.WebJobs.Host/Bindings/BindingProviders/BindToStreamBindingProvider.cs

```diff
@@ -0,0 +1,501 @@
+            prop.SetValue(attribute, access); 
+        }
+                
+        // As a binding, this is one per parameter, shared across each invocation instance.
+        private class ReadExactBinding : BindingBase<TAttribute>
+        {
+            private readonly BindToStreamBindingProvider<TAttribute> _parent;
+            private readonly Type _userType;
+            private readonly FileAccess _targetFileAccess;
+            private readonly Type _typeValueProvider;
+
+            public ReadExactBinding(
+                AttributeCloner<TAttribute> cloner,
+                ParameterDescriptor param,
+                BindToStreamBindingProvider<TAttribute> parent,
+                Type argHelper,
+                Type userType,
+                FileAccess targetFileAccess)
+                : base(cloner, param)
+            {
+                _parent = parent;
+                _userType = userType;
+                _targetFileAccess = targetFileAccess;
+                _typeValueProvider = argHelper;
+            }
+
+            protected override async Task<IValueProvider> BuildAsync(TAttribute attrResolved, ValueBindingContext context)
+            {
+                // set FileAccess beofre calling into the converter. Don't want converters to need to deal with a null FileAccess.
+                SetFileAccessFromAttribute(attrResolved, _targetFileAccess);
+
+                var patternMatcher = _parent._patternMatcher;
+                Func<object, object> builder = patternMatcher.TryGetConverterFunc(typeof(TAttribute), typeof(Stream));
+                Func<Stream> builder2 = () => (Stream)builder(attrResolved);
+
+
+                BaseValueProvider valueProvider = (BaseValueProvider)Activator.CreateInstance(_typeValueProvider);
+                await valueProvider.InitAsync(builder2, _userType);
+             
+                return valueProvider;
+            }
+        }
+
+        // The base IValueProvider. Handed  out per-instance
+        // This wraps the stream and coerces it to the user's parameter.
+        private abstract class BaseValueProvider : IValueBinder
+        {
+            private Stream _stream;
+            public Type Type { get; set; } // Impl IValueBinder
+
+            private string _invokeString;
+
+            // Helper to build the stream. This will only get invoked once and then cached as _stream. 
+            private Func<Stream> _streamBuilder;
+
+            object _userArg;
+
+            protected Stream GetOrCreateStream()
+            {
+                if (_stream == null)
+                {
+                    _stream = _streamBuilder();
+                }
+                return _stream;
+            }
+
+            public async Task InitAsync(Func<Stream> builder, Type userType)
+            {
+                Type = userType;
+                _invokeString = "???";
+                _streamBuilder = builder;
+
+                _userArg = await this.CreateUserArgAsync();
```

Comment:
The "user arg" naming isn't clear. Can you rename things to `_value = await GetValueAsync();`?

Reply:
I'll do _userValue. by 'user', I mean value passed to the user's function (as opposed to the internal values we track) 

---
In reply to: [145444786](https://github.com/Azure/azure-webjobs-sdk/pull/1387#discussion_r145444786) [](ancestors = 145444786)